### PR TITLE
Md5 module

### PR DESCRIPTION
**Pull request by nicwest**
Hey guys I wrote this for largely to see if I could, I can appreciate there are probably more straight forward ways of going about md5-ing a blob of text in vim.

I followed the [wikipedia pseudo code for the algo](https://en.wikipedia.org/wiki/MD5#Algorithm) and the tests use the test hashes from [RFC 1321](https://tools.ietf.org/html/rfc1321#appendix-A.5)

I've added a `Hash` module as I was thinking about doing a couple of the more interesting algorithms and adding something along the lines of python's `hashlib` https://docs.python.org/3.6/library/hashlib.html

Let me know what you think!

### CODE REVIEW

**Comment by nicwest** [comment]
I can reproduce the errors shown by CI with this VagrantFile:
```
# -*- mode: ruby -*-
# vi: set ft=ruby :
Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu/trusty32"
end
```

Looks to be a 32 bit issue, will debug this evening.

File: autoload/vital/__vital__/Hash/MD5.vim [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+function! s:_int2bytes(bits, int) abort
+  return map(range(a:bits / 8), 's:bitwise.and(s:bitwise.rshift(a:int, v:val * 8), 0xff)')
+endfunction
+
+function! s:_bytes2int32(bytes) abort
+  return  or(s:bitwise.lshift(a:bytes[3], 24), 
+        \ or(s:bitwise.lshift(a:bytes[2], 16),
+        \ or(s:bitwise.lshift(a:bytes[1], 8),
+        \ a:bytes[0])))
+endfunc
```
Comment (tyru) [inline autoload/vital/__vital__/Hash/MD5.vim]: please use `:endfunction`

File: test/Hash/MD5.vim [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+function! s:suite.before()
+  let s:MD5 = vital#vital#new().import('Hash.MD5')
+endfunction
+
+function! s:suite.after()
+  unlet! s:MD5
+endfunction
+
+func! s:suite.encode() abort
```
Comment (tyru) [inline test/Hash/MD5.vim]: please use `:function!`

**Comment by rhysd** [comment]
I'm very sorry for late response. I actually did not check PRs frequently and missed this (and I'm surprised no one did not review this PR).

If you feel ok, please check my review comments

File: doc/vital/Hash/MD5.txt [hunk 3]
```diff
@@ -0,0 +1,27 @@
+*vital/Hash/MD5.txt*		md5 utilities library.
```
Comment (rhysd) [inline doc/vital/Hash/MD5.txt]: Please update tables in `README.md` and `doc/vital.txt` by adding this module.

File: test/Hash/MD5.vim (hunk 2 above)
Comment (rhysd) [inline test/Hash/MD5.vim]: Please use non-abbreviated names persistently (here `function!` should be used)

**Comment by tyru** [comment]
@nicwest Sorry for delay. I just reviewed your changes.
but I don't know about implementation of MD5 algorithm.
I hope someone else (or @rhysd) review this PR too...

**Comment by tyru** [comment]
@rhysd Do you submit your comments? Your comments are not in diff.

**Comment by rhysd** [comment]
I'm on an express train 🚄  now and network condition is poor. So some review comments were reflected later...

I'm not familiar with MD5 algorithm, but the implementation seems ok as long as comparing with C implementation and some tests were added from RFC.

**Comment by rhysd** [comment]
@nicwest Broken CI issue was resolved at https://github.com/vim-jp/vital.vim/pull/543. Could you merge or rebase the latest master branch?

**Comment by rhysd** [comment]
@nicwest Thank you for this PR. I'm sorry for making you wait for a long time.

**Pull request merged**